modules/videoio/src/cap_mjpeg_decoder.cpp

```diff
@@ -40,13 +40,733 @@
+        return (uint32_t)m_mjpeg_frames.size();
+
+    return (uint32_t)(m_frame_iterator - m_mjpeg_frames.begin() + 1);
+}
+
+/*
+     -   **CV_CAP_PROP_FRAME_WIDTH** Width of the frames in the video stream.
+     -   **CV_CAP_PROP_FRAME_HEIGHT** Height of the frames in the video stream.
+     -   **CV_CAP_PROP_FPS** Frame rate.
+     -   **CV_CAP_PROP_FOURCC** 4-character code of codec.
+     -   **CV_CAP_PROP_FRAME_COUNT** Number of frames in the video file.
+     -   **CV_CAP_PROP_FORMAT** Format of the Mat objects returned by retrieve() .
+     -   **CV_CAP_PROP_MODE** Backend-specific value indicating the current capture mode.
+*/
+
+bool MotionJpegCapture::setProperty(int property, double value)
+{
+    if(property == CAP_PROP_POS_FRAMES)
+    {
+        if(int(value) == 0)
+        {
+            m_is_first_frame = true;
+            m_frame_iterator = m_mjpeg_frames.end();
+            return true;
+        }
+        else if(m_mjpeg_frames.size() > value)
+        {
+            m_frame_iterator = m_mjpeg_frames.begin() + int(value - 1);
+            m_is_first_frame = false;
+            return true;
+        }
+    }
+
+    return false;
+}
+
+double MotionJpegCapture::getProperty(int property) const
+{
+    switch(property)
+    {
+        case CAP_PROP_POS_FRAMES:
+            return (double)getFramePos();
+        case CAP_PROP_POS_AVI_RATIO:
+            return double(getFramePos())/m_mjpeg_frames.size();
+        case CAP_PROP_FRAME_WIDTH:
+            return 0;
+        case CAP_PROP_FRAME_HEIGHT:
+            return 0;
+        case CAP_PROP_FPS:
+            return 0;
+        case CAP_PROP_FOURCC:
+            return (double)CV_FOURCC('M','J','P','G');
+        case CAP_PROP_FRAME_COUNT:
+            return (double)m_mjpeg_frames.size();
+        case CAP_PROP_FORMAT:
+            return 0;
+        default:
+            return 0;
+    }
+}
+
+std::vector<char> MotionJpegCapture::readFrame(frame_iterator it)
+{
+    m_file_stream.seekg(m_movi_pos + it->first);
+
+    RiffChunk chunk;
+    m_file_stream >> chunk;
+
+    std::vector<char> result;
+
+    result.reserve(chunk.m_size);
+    result.resize(chunk.m_size);
+
+    m_file_stream.read(result.data(), chunk.m_size);
+
+    return result;
+}
+
+bool MotionJpegCapture::grabFrame()
+{
+    if(isOpened())
+    {
+        if(m_is_first_frame)
+        {
+            m_is_first_frame = false;
+            m_frame_iterator = m_mjpeg_frames.begin();
+        }
+        else
+        {
+            ++m_frame_iterator;
+        }
+    }
+
+    return m_frame_iterator != m_mjpeg_frames.end();
+}
+
+bool MotionJpegCapture::retrieveFrame(int, OutputArray output_frame)
+{
+    if(m_frame_iterator != m_mjpeg_frames.end())
+    {
+        std::vector<char> data = readFrame(m_frame_iterator);
+
+        if(data.size())
+        {
+            m_current_frame = imdecode(data, CV_LOAD_IMAGE_ANYDEPTH | CV_LOAD_IMAGE_COLOR);
+        }
+
+        output_frame.getMatRef() = m_current_frame;
```

Comment:
it may be not a good solution. Please, use m_current_frame.copyTo(output_frame) instead.